### ISSUE TITLE
Wrong error message when capturing const object by reference

### ISSUE THREAD

**Issue by vovchikzd** [llvm/llvm-project#98772]
When trying to create a mutable lambda with const object captured by reference and then mutate that object 
``` c++
int main() {
    const int a = 69;
    auto lambda = [&r = a](int x) mutable {
        r += x;
        return r;
    };
}
```
you got incorrect error message
```
<source>:4:11: error: cannot assign to a variable captured by copy in a non-mutable lambda
    4 |         r += x;
      |         ~ ^
```
but r is captured by reference and lambda is mutable
godbolt example: https://godbolt.org/z/sKd3bh6aM
